File: microservices/frontend/src/circuitbreaker.js
class CircuitBreaker {
  constructor(maxFailures = 3, resetTimeoutMs = 30000) {
    this.maxFailures = maxFailures;
    this.resetTimeout = resetTimeoutMs;
    this.state = 'CLOSED'; // CLOSED, OPEN, HALF_OPEN
    this.failures = 0;
    this.lastFailure = null;
  }

  async call(fn) {
    if (this.state === 'OPEN') {
      if (Date.now() - this.lastFailure > this.resetTimeout) {
        this.state = 'HALF_OPEN';
        this.failures = 0;
      } else {
        throw new Error('Circuit breaker is open - authentication service unavailable');
      }
    }

    try {
      const result = await fn();
      
      // Success
      if (this.state === 'HALF_OPEN') {
        this.state = 'CLOSED';
      }
      this.failures = 0;
      return result;
    } catch (error) {
      this.failures++;
      this.lastFailure = Date.now();
      
      if (this.failures >= this.maxFailures) {
        this.state = 'OPEN';
      }
      throw error;
    }
  }

  getState() {
    return this.state;
  }

  getFailures() {
    return this.failures;
  }

  isOpen() {
    return this.state === 'OPEN';
  }

  isClosed() {
    return this.state === 'CLOSED';
  }

  isHalfOpen() {
    return this.state === 'HALF_OPEN';
  }
}

export default CircuitBreaker;

File: microservices/auth-api/circuitbreaker.go
package main

import (
	"errors"
	"sync"
	"time"
)

type CircuitBreakerState int

const (
	Closed CircuitBreakerState = iota
	Open
	HalfOpen
)

type CircuitBreaker struct {
	maxFailures  int
	resetTimeout time.Duration
	state        CircuitBreakerState
	failures     int
	lastFailure  time.Time
	mutex        sync.RWMutex
}

func NewCircuitBreaker(maxFailures int, resetTimeout time.Duration) *CircuitBreaker {
	return &CircuitBreaker{
		maxFailures:  maxFailures,
		resetTimeout: resetTimeout,
		state:        Closed,
	}
}

func (cb *CircuitBreaker) Call(fn func() error) error {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	if cb.state == Open {
		if time.Since(cb.lastFailure) > cb.resetTimeout {
			cb.state = HalfOpen
			cb.failures = 0
		} else {
			return errors.New("circuit breaker is open - users service unavailable")
		}
	}

	err := fn()
	
	if err != nil {
		cb.failures++
		cb.lastFailure = time.Now()
		
		if cb.failures >= cb.maxFailures {
			cb.state = Open
		}
		return err
	}

	// Success
	if cb.state == HalfOpen {
		cb.state = Closed
	}
	cb.failures = 0
	return nil
}

func (cb *CircuitBreaker) GetState() CircuitBreakerState {
	cb.mutex.RLock()
	defer cb.mutex.RUnlock()
	return cb.state
}

func (cb *CircuitBreaker) GetFailures() int {
	cb.mutex.RLock()
	defer cb.mutex.RUnlock()
	return cb.failures
}